solutions/cf-718C.mdx

```diff
@@ -0,0 +1,176 @@
+---
+id: cf-718C
+source: cf
+title: Sasha and Array
+author: Dustin Miao
+---
+
+## Explanation
+
+Firstly, a note on notation: let $F_i$ denote the $i$th fibonacci number. That is,
+
+$$
+F_i = 
+\begin{cases} 
+	0 & i = 0 \\ 
+	1 & i = 1 \\
+	F_{i - 1} + F{i - 2} & i \geq 2 \\
+\end{cases}
+$$
+
+In this problem, we will build a segment tree over the array. For a leaf node in the segment tree with value $v$, we will store a pair of values $(F_{v - 1}, F_v)$. For all non-leaf nodes, we will store a pair of values equal to the pair-sum of its children. That is, if a node $u$ has children $v$ and $w$, then $$u = (v_0 + w_0, v_1 + w_1)$$. 
+
+We will use the term _cycle by k_ to denote transforming some pair $(F_{i - 1}, F_i)$ to $(F_{i - 1 + k}, F_{i + k})$. By default, the term _cycle_ refers to _cycle by 1_.
+
+For each update, we need to cycle each leaf node by some amount $x$ such that the value in a node affected goes from $(F_{v - 1}, F_v)$ to $(F_{v - 1 + x}, F_{v + x})$. We can do this using matrix exponentiation.
+
+First note that for a matrix $\begin{pmatrix} F_{i - 1} & F_i \end{pmatrix}$, we can cycle it by multiplying by $\begin{pmatrix} 0 & 1 \\ 1 & 1 \end{pmatrix}$. That is,
+
+$$
+\begin{pmatrix} F_i & F_{i + 1} \end{pmatrix} = \begin{pmatrix} F_{i - 1} & F_i \end{pmatrix}\begin{pmatrix} 0 & 1 \\ 1 & 1 \end{pmatrix}
+$$
+
+Inductively, we can cycle by $k$ by multiplying multiple times:
+
+$$
+\begin{pmatrix} F_{i - 1 + k} & F_{i + k} \end{pmatrix} = \begin{pmatrix} F_{i - 1} & F_i \end{pmatrix}\begin{pmatrix} 0 & 1 \\ 1 & 1 \end{pmatrix}^k
+$$
+
+We can use binary exponentiation to quickly compute powers of the two-by-two matrix. 
+
+Secondly, we need to utilize the distribute property of matrix multiplication to same-size matricies. That is, if $m_k$ is a one-by-two matrix for $k \in [1, n]$, then the following property is satisfied:
+
+$$
+(m_1 + m_2 + \dots + m_n) \cdot \begin{pmatrix} 0 & 1 \\ 1 & 1 \end{pmatrix} = m_1 \cdot \begin{pmatrix} 0 & 1 \\ 1 & 1 \end{pmatrix} + m_2 \cdot \begin{pmatrix} 0 & 1 \\ 1 & 1 \end{pmatrix} + \dots + m_n \cdot \begin{pmatrix} 0 & 1 \\ 1 & 1 \end{pmatrix}
+$$
+
+In that way, we can utilize lazy propgation on the segment tree. We will store an integer tag in each node denoting the lazy update, which is denotes that we must cycle every node in its subtre by the value of the tag. When propogating, we can simply update the node by binary exponentiating and then multiplying, then storing it in the lazy tag. Querying works as normal on the segment tree. 
+
+## Implementation
+
+<LanguageSection>
+
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+#define FF first
+#define SS second
```

Comment:
use only the crap here
https://github.com/cpinitiative/usaco-guide/issues/806

Reply:
well good thing this isn't bronze through gold then

Reply:
I hate you so much.
![image](https://user-images.githubusercontent.com/55369003/143468481-d272024c-072b-4791-b033-360ff2b59408.png)

\lh